[quant] Speed up per-channel min-max observer

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34118 [quant] Speed up per-channel min-max observer**

Summary: Previously calc_per_channel_qparams was using for loops and python primitives, which called `item` many times causing slowdown during training.
    These changes uses torch primitives on the tensor to speed up the operation over 60x

    Perf results on MobileNetV2 during training using autograd profiler

    FP32 forward call -
    Self CPU time total: 47.222ms
    CUDA time total: 124.001ms

    before change
    FakeQuant Model -
    Self CPU time total: 19.107s
    CUDA time total: 27.177s

    after change
    FakeQuant Model -
    Self CPU time total: 404.667ms
    CUDA time total: 446.344ms

Test Plan:
    python test/test_quantization.py


Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20287841](https://our.internmc.facebook.com/intern/diff/D20287841)

supriyar committed:
[quant] Speed up per-channel min-max observer
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Speed up per-channel min-max observer"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Speed up per-channel min-max observer"


Summary: Previously calc_per_channel_qparams was using for loops and python primitives, which called `item` many times causing slowdown during training.
    These changes uses torch primitives on the tensor to speed up the operation over 60x

    Perf results on MobileNetV2 during training using autograd profiler

    FP32 forward call -
    Self CPU time total: 47.222ms
    CUDA time total: 124.001ms

    before change
    FakeQuant Model -
    Self CPU time total: 19.107s
    CUDA time total: 27.177s

    after change
    FakeQuant Model -
    Self CPU time total: 404.667ms
    CUDA time total: 446.344ms

Test Plan:
    python test/test_quantization.py


Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -144,20 +144,49 @@ def _calculate_per_channel_qparams(self, min_vals, max_vals):
-                min_vals[i] <= max_vals[i]
-            ), "min {} should be less than max {}".format(min_vals[i], max_vals[i])
+        diff = min_vals <= max_vals
+        for i in range(len(diff)):
```
Can you instead do assert(sum(diff) ==len(diff))?

raghuramank100 reviewed (approved):
Looks good!

supriyar committed:
Update on "[quant] Speed up per-channel min-max observer"


Summary: Previously calc_per_channel_qparams was using for loops and python primitives, which called `item` many times causing slowdown during training.
    These changes uses torch primitives on the tensor to speed up the operation over 60x

    Perf results on MobileNetV2 during training using autograd profiler

    FP32 forward call -
    Self CPU time total: 47.222ms
    CUDA time total: 124.001ms

    before change
    FakeQuant Model -
    Self CPU time total: 19.107s
    CUDA time total: 27.177s

    after change
    FakeQuant Model -
    Self CPU time total: 404.667ms
    CUDA time total: 446.344ms

Test Plan:
    python test/test_quantization.py


Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
